php/src/Google/Protobuf/Internal/DescriptorProto.php

```diff
@@ -307,7 +307,7 @@ public function getOptions()
      * @param \Google\Protobuf\Internal\MessageOptions $var
      * @return $this
      */
-    public function setOptions($var)
+    public function setOptions(\Google\Protobuf\Internal\MessageOptions $var = null)
```

Comment:
I'm excited to see this change, but should we consider this breaking? If users pass in anything other than `MessageOptions|null` they will get a different error now than the one provided by `GPBUtil::checkMessage`. 

Assuming we move forward, however, a few thoughts:

- With the typecheck in place, are there any objections to removing the `GPBUtil` check?
- There is a mixture of style in the class, so I understand not being strict about this - but if we are open to the idea, moving the needle towards using relative namespaces instead of absolute would be great :).

Reply:
What do you mean by relative namespace? Could you give me an example?

Reply:
> Currently, whether to throw error or exception for getter/setter/constructor is undefined behavior.

I'm not sure I understand. Without this changeset it is currently possible for users to have code like the following:

```php
use Google\Protobuf\Internal\DescriptorProto;

try {
    (new DescriptorProto)
        ->setOptions(new \stdClass);
} catch (\Exception $ex) {
    // Handling the exception thrown by GPBUtil::checkMessage
    echo 'There was a problem.';
}
```

On PHP 7+, for example, with this changeset added the above code would break with the following `TypeError`:

```
PHP Fatal error:  Uncaught TypeError: Argument 1 passed to
Google\Protobuf\Internal\DescriptorProto::setOptions() must
be an instance of Google\Protobuf\Internal\MessageOptions
or null, instance of stdClass given
```

> GPBUtil also checks other primitive fields. Since php5 doesn't provide typehint for primitive fields and we are currently still supporting that. We cannot remove GPBUtil.

Yes, that definitely is fair for primitives. I didn't mean to suggest removing the check across the board, just for situations where we would already be relying on a typehint.

> What do you mean by relative namespace? Could you give me an example?

Sure!

```php
// Instead of..
public function setOptions(\Google\Protobuf\Internal\MessageOptions $var = null)
{ }

// Import the class and use the relative name
use Google\Protobuf\Internal\MessageOptions;

public function setOptions(MessageOptions $var = null)
{ }
```

Reply:
Expecting setter to throw an exception is an undefined behavior. Users should not write code like that. So I think we are good to move forward.

Reply:
I think it's ok to remove the message check part in GPBUtil.

Reply:
What if there are same message name under different namespace? e.g., A\Foo and B\Foo

Reply:
I still don't understand what you mean by undefined behavior. We have GA level libraries with these generated classes living in our codebase over on google-cloud-php. We have no such notice published anywhere so I don't know how users would know they shouldn't catch the exception that is thrown by the validation utility.

> What if there are same message name under different namespace? e.g., A\Foo and B\Foo

Good question. In cases like that aliases can be used:

```php
use A\Foo as AFoo;
use B\Foo;
```

Reply:
By "undefined behavior", I mean although the code works like this, but we have never said we will guarantee such behavior. Such usage of undefined behavior may be broken by future release. (I think it's a common situation for opensource library).

Reply:
ummm... I feel it's a little bit intuitive from AFoo to figure out A\Foo.
What if both A\Foo and AFoo already exist?

Reply:
To state all undefined behavior is impracticable. Unless there is unlimited test cases, we cannot know how many undefined behaviors there are. The only way we can provide is unless we explicitly say "You can rely on this", users should not do that.

Reply:
> To state all undefined behavior is impracticable.

Yes, agreed. That is why it should be an edge case.

Could you help me understand what process we use here to define behavior as "expected"?

In my experience, once we have a stable public API the "expected" behavior is the exact behavior that the code performs and users should be able to rely on that. The exception being bugs and/or when we document that there may be an outlier for a given reason (for example, perhaps we mark a publicly accessible method as internal and say it shouldn't be relied on by users).

Reply:
https://developers.google.com/protocol-buffers/docs/reference/php-generated
We have public doc for the API. The expected behavior should be stated there.
Although I agree this doc may miss some API later added. We should make the doc complete.

Reply:
Even if we attempt to complete the document, there is a very high chance it will fall out of alignment as new features are introduced. It seems very risky to ask users to rely on that as the source of truth for how they can expect the library to function.

I do really like this change, but would rather we defer it to the next major version if possible.